ocaml/testsuite/tests/typing-unique/overwriting.ml

```diff
@@ -661,6 +787,169 @@ Error: Overwrite may not change the tag to OptionA.
 Hint: The old tag of this allocation is unknown.
 |}]
 
+type 'a mutable_record = { mutable m : 'a }
+
+let mutable_field_aliased r =
+  let y = OptionA "foo" in
+  r.m <- y;
+  (unique_ r), y
+[%%expect{|
+type 'a mutable_record = { mutable m : 'a; }
+val mutable_field_aliased :
+  unique_ options mutable_record -> options mutable_record * options @@
+  global many = <fun>
+|}]
+
+let mutable_field_aliased r =
+  unique_ r.m
+[%%expect{|
+val mutable_field_aliased : unique_ 'a mutable_record -> 'a @@ global many =
+  <fun>
+|}]
+
+let tag_of_mutable_field r =
+  match r with
+  | { m = OptionA s } ->
+    overwrite_ r.m with OptionA s
+  | _ -> OptionB ""
+[%%expect{|
+Line 4, characters 4-33:
+4 |     overwrite_ r.m with OptionA s
+        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Alert Translcore: Overwrite not implemented.
+Uncaught exception: File "ocaml/parsing/location.ml", line 1106, characters 2-8: Assertion failed
+
+|}]
+
+let mutating_tag_seq r =
+  match r with
+  | { m = OptionA s } ->
+    r.m <- OptionB s;
+    overwrite_ r.m with OptionA s
+  | _ -> OptionB ""
+[%%expect{|
+Line 5, characters 24-31:
+5 |     overwrite_ r.m with OptionA s
+                            ^^^^^^^
+Error: Overwrite may not change the tag to OptionA.
+Hint: The old tag of this allocation was changed through mutation.
+|}]
+
+let mutating_tag_seq_parent r =
+  match r with
+  | { m = { x = OptionA s } } ->
+    r.m <- { x = OptionB s };
+    overwrite_ r.m.x with OptionA s
+  | _ -> OptionB ""
+[%%expect{|
+Line 5, characters 26-33:
+5 |     overwrite_ r.m.x with OptionA s
+                              ^^^^^^^
+Error: Overwrite may not change the tag to OptionA.
+Hint: The old tag of this allocation was changed through mutation.
+|}]
+
+let mutating_tag_par r =
+  match r with
+  | { m = OptionA s } ->
+    (r.m <- OptionB s), overwrite_ r.m with OptionA s
+  | _ -> (), OptionB ""
+[%%expect{|
+Line 4, characters 44-51:
+4 |     (r.m <- OptionB s), overwrite_ r.m with OptionA s
+                                                ^^^^^^^
+Error: Overwrite may not change the tag to OptionA.
+Hint: The old tag of this allocation was changed through mutation.
```

Comment:
Well, actually, expressions in tuples evaluate right-to-left. (Strange but true.) So in practice this would be fine. We probably shouldn't allow it, though. But if we don't, the error message isn't great, because of the "was".